azext_edge/edge/providers/orchestration/resource_map.py

```diff
@@ -0,0 +1,171 @@
+                    )
+                )
+        return result
+
+    @property
+    def custom_locations(self) -> List[IoTOperationsResource]:
+        result = []
+        if "customLocations" in self._resource_map and self._resource_map["customLocations"]:
+            for cl_id in self._resource_map["customLocations"]:
+                result.append(
+                    IoTOperationsResource(
+                        resource_id=cl_id,
+                        display_name=self._resource_map["customLocations"][cl_id]["name"],
+                        api_version=self._resource_map["customLocations"][cl_id]["apiVersion"],
+                    )
+                )
+        return result
+
+    def get_resource_sync_rules(self, custom_location_id: str) -> List[IoTOperationsResource]:
+        result = []
+        if (
+            "customLocations" in self._resource_map
+            and self._resource_map["customLocations"]
+            and custom_location_id in self._resource_map["customLocations"]
+            and self._resource_map["customLocations"][custom_location_id]
+            and "resourceSyncRules" in self._resource_map["customLocations"][custom_location_id]
+            and self._resource_map["customLocations"][custom_location_id]["resourceSyncRules"]
+        ):
+            for rsr in self._resource_map["customLocations"][custom_location_id]["resourceSyncRules"]:
+                result.append(
+                    IoTOperationsResource(
+                        resource_id=rsr["id"], display_name=rsr["name"], api_version=rsr["apiVersion"]
+                    )
+                )
+        return result
+
+    def get_resources(self, custom_location_id: str) -> List[IoTOperationsResource]:
+        result: List[IoTOperationsResource] = []
+        if (
+            "customLocations" in self._resource_map
+            and self._resource_map["customLocations"]
+            and custom_location_id in self._resource_map["customLocations"]
+            and self._resource_map["customLocations"][custom_location_id]
+            and "resources" in self._resource_map["customLocations"][custom_location_id]
+            and self._resource_map["customLocations"][custom_location_id]["resources"]
+        ):
+            for resource in self._resource_map["customLocations"][custom_location_id]["resources"]:
+                result.append(
+                    IoTOperationsResource(
+                        resource_id=resource["id"],
+                        display_name=resource["name"],
+                        api_version=resource["apiVersion"],
+                        segments=len(resource["id"].split("/")),
+                    )
+                )
+
+            result = sorted(result, key=lambda r: (r.segments, r.display_name.lower()), reverse=True)
+
+        return result
+
+    def refresh_related_resource_ids(
+        self,
+    ) -> Dict[str, Union[List[Dict[str, str]], Dict[str, Dict[str, Union[List[Dict[str, str]], str]]]]]:
```

Comment:
nit: this is a very scary typing

Reply:
would be more helpful to have the typing of each resource (what type is the stored extension, etc)

Reply:
I re-wrote this module that includes simplifying this typing.